src/Components/Server/src/Circuits/CircuitPrerenderer.cs

```diff
@@ -2,47 +2,74 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
+using System.Linq;
 using System.Runtime.ExceptionServices;
 using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Server.Prerendering;
 using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Http.Extensions;
 
 namespace Microsoft.AspNetCore.Components.Server.Circuits
 {
     internal class CircuitPrerenderer : IComponentPrerenderer
     {
+        private static object CircuitHostKey = new object();
+
         private readonly CircuitFactory _circuitFactory;
+        private readonly CircuitRegistry _registry;
 
-        public CircuitPrerenderer(CircuitFactory circuitFactory)
+        public CircuitPrerenderer(CircuitFactory circuitFactory, CircuitRegistry registry)
         {
             _circuitFactory = circuitFactory;
+            _registry = registry;
         }
 
-        public async Task<IEnumerable<string>> PrerenderComponentAsync(ComponentPrerenderingContext prerenderingContext)
+        public async Task<ComponentPrerenderResult> PrerenderComponentAsync(ComponentPrerenderingContext prerenderingContext)
         {
             var context = prerenderingContext.Context;
-            var circuitHost = _circuitFactory.CreateCircuitHost(
-                context,
-                client: CircuitClientProxy.OfflineClient,
-                GetFullUri(context.Request),
-                GetFullBaseUri(context.Request));
+            var circuitHost = GetOrCreateCircuitHost(context);
 
-            // We don't need to unsubscribe because the circuit host object is scoped to this call.
-            circuitHost.UnhandledException += CircuitHost_UnhandledException;
+            var renderResult = await circuitHost.PrerenderComponentAsync(
+                prerenderingContext.ComponentType,
+                prerenderingContext.Parameters);
 
-            // For right now we just do prerendering and dispose the circuit. In the future we will keep the circuit around and
-            // reconnect to it from the ComponentsHub. If we keep the circuit/renderer we also need to unsubscribe this error
-            // handler.
-            try
+            circuitHost.Descriptors.Add(new ComponentDescriptor
             {
-                return await circuitHost.PrerenderComponentAsync(
-                    prerenderingContext.ComponentType,
-                    prerenderingContext.Parameters);
+                ComponentType = prerenderingContext.ComponentType,
+                Selector = $"[data-component-id={renderResult.ComponentId}]",
+                Prerendered = true
+            });
+
+            var result = new[] { $"<div data-circuit-id=\"{circuitHost.CircuitId}\" data-renderer-id=\"{circuitHost.Renderer.Id}\" data-component-id=\"{renderResult.ComponentId}\">" }
```

Comment:
I understand this will change

Reply:
Yes. For this PR i'm leaving it like this, but I plan to move it to use comments ASAP.

Reply:
I ended up removing it.